### DISCUSSION TITLE
Adding quote blocks and colored text using notional

### DISCUSSION THREAD

**Discussion by HaiderHB** [jheddings/notional#128] [Q&A]
Hey,

I wanted to use the notional api for some code that adds my kindle clippings to a notion page. Currently here is how its done:

`
            new_page = notion.pages.create(
            parent=notion.databases.retrieve(notion_database_id),

            properties={

                "Title": Title[title],

                "Author": RichText[author],

                "Highlights": Number[clippings_count],

                "Last Highlighted": Date[last_date.isoformat()],

                "Last Synced": Date[datetime.now().isoformat()],

            },

            children=[],
        )
        page_content = Paragraph["".join(formatted_clippings)]
        notion.blocks.children.append(new_page, page_content)`

here "formatted_clippings" is a list of all clippings.

I just wanted to ask, how can I make it so the clippings being added are added in quote blocks instead?

My guess is ill have to edit the line: page_content = Paragraph["".join(formatted_clippings)]
and also use: from notional.blocks import Quote

but im not too sure how to go about it.

**Comment by jheddings** [comment]
Hi @SHaiderX !

You are correct...  There is a `Quote` block type that you can use:

```
from notional import blocks

page_content = blocks.Quote[formatted_clippings]
```

The compose interface (square brackets) allow you to provide a list of strings that will be appended as text object in Notion.  Give that a shot and check back with any issues.

**Reply by HaiderHB** [reply]
Oh I see, thanks alot! Also, is there any way to change the text color? Like lets say I want to make the quote in the color yellow, how would I do that?

**Comment by jheddings** [comment]
Notion uses an annotated `rich_text` object for handling display attributes of text.  When you generate your text, you'll need to use this object rather than plain strings to set the formatting:

```python
from notional.blocks import Quote
from notional.text import Annotations, FullColor, RichTextObject

comments = []

# add comments as RichTextObjects with yellow background
highlight = Annotations(color=FullColor.YELLOW_BACKGROUND)
text = "this is very interesting!"
rtf = TextObject[text, None, highlight]
comments.append(rtf)

# create a Quote block from the comments
quote = Quote[comments]

# append the quote block to our page
notion.blocks.children.append(my_page, quotes)
```

This gets to be challenging if you want to mix plan and styled text, since you need to break it into multiple `RichTextObject`'s, some with style and some without.  For example:

```python
highlight = Annotations(color=FullColor.YELLOW_BACKGROUND)
bold = Annotations(bold=True)

comments.append(
    TextObject["this", None, highlight],
    TextObject[" is "],
    TextObject["very", None, bold],
    TextObject[" interesting!"],
)

...
```

(BTW - I have not tested the above code, so hopefully it is close to what you need!)

**Comment by jheddings** [comment]
Also, depending on your use case, check out the `orm.py` example for expressing your database pages as a Python object.  It makes managing the properties and creating new pages much easier!

**Comment by HaiderHB** [comment]
Thanks alot for the suggestion!
Im currently trying to figure out how to get this to work:
![image](https://user-images.githubusercontent.com/67211098/216716380-902c4cac-066d-4ded-8f86-734c5f2f199a.png)

Do you have any idea what I should do here?

**Reply by jheddings** [reply]
Ah, yeah...  You'll just have to use positional arguments for the compose interface:

```python
# parameters: text, href, style
rtf = TextObject[text, None, highlight]
```

(Updated the notes above for completeness)

**Comment by HaiderHB** [comment]
Sorry its actually still giving me an error. In these lines:


            # add comments as RichTextObjects with yellow background
            highlight = Annotations(color=FullColor.YELLOW_BACKGROUND)
            text = "this is very interesting!"
            rtf = RichTextObject[text, None, highlight]
            formatted_clippings.append(Quote[rtf])
            print(rtf)

when I print rtf it is printing None. The quote blocks being added are empty, they dont have the text, let alone the color

**Reply by HaiderHB** [reply]
I tried it inside the comment = [] list too, was still doing nothing

**Reply by jheddings** [reply]
Ah, you are right!  I’m at a computer to try this out and there is a bug in the `RichTextObject`…  Your code looks correct.  I will get a patch going to fix the problem.

As a workaround, you can use the constructor directly:

```python
rtf = TextObject(plain_text=text, style=highlight)
```

Sorry for the trouble!
-jh

**Reply by jheddings** [reply]
Okay, notional-0.7.3 should fix the bug for you to use the compose interface if you prefer.  Sorry about the churn!

**Comment by HaiderHB** [comment]
Thanks alot for your help. And I feel super bad bugging you again and again but there is one more error im getting and im not sure what im doing wrong:

        # add comments as RichTextObjects with yellow background
        highlight = Annotations(color=FullColor.YELLOW_BACKGROUND)
        text = "Page: " + page
        rtf = RichTextObject[text, None, highlight]
        formatted_clippings.append(Quote[rtf])

        for clip in formatted_clippings:
            notion.blocks.children.append(new_page, clip)

When I do this, the final line gives the error:
body.children[0].quote.rich_text[0].text should be defined, instead was `undefined`.
body.children[0].quote.rich_text[0].mention should be defined, instead was `undefined`.
body.children[0].quote.rich_text[0].equation should be defined, instead was `undefined`.

Before when formatted_clippings just had quote blocks it was working fine, but now that I added the highlights to it it gives me that.

**Reply by jheddings** [reply]
In this case, `RichTextObject` is a base type for other objects...  I'm guessing you want to use `TextObject` here instead (rich text can also be a Mention or Equation object).

So for your example, try:

```python
    # add comments as RichTextObjects with yellow background
    highlight = Annotations(color=FullColor.YELLOW_BACKGROUND)
    text = "Page: " + page
    rtf = TextObject[text, None, highlight]
    formatted_clippings.append(Quote[rtf])

    for clip in formatted_clippings:
        notion.blocks.children.append(new_page, clip)
```

And it's no bother...  I'm glad to see people using the library!

**Reply by HaiderHB** [reply]
Thanks alot! it works now :)
